### DISCUSSION TITLE
Welcome to midi_degradation_toolkit Discussions!

### DISCUSSION THREAD

**Discussion by JamesOwers** [JamesOwers/midi_degradation_toolkit#169] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MarcusLee-awsome** [comment]
I want to use degrader.py in my model as midi augmentation, but I don't know how to convert midi
file into pd.dataFrame, and then convert it to midi back into my model

**Reply by JamesOwers** [reply]
Hi there. You should be able to use fileio.midi_to_df to do this conversion for you, then df_to_midi to go back.

https://github.com/JamesOwers/midi_degradation_toolkit/blob/master/mdtk/fileio.py#L115
https://github.com/JamesOwers/midi_degradation_toolkit/blob/f827c1b6dc3cb51349fb1f6b5206152aed8c21c6/mdtk/fileio.py#L309

**Reply by MarcusLee-awsome** [reply]
thank you so much

**Reply by MarcusLee-awsome** [reply]
I try to use the fileio.py to generate a new midi file, although the code can go smoothly, the file can not be generated. Can you give me some advice?

**Reply by JamesOwers** [reply]
I can try! Please share a minimal working code example showing what you’re trying to do which I can try to run.

**Reply by MarcusLee-awsome** [reply]
sorry，let you wait for a long time. I have a job interview today , and I can't git my code to my account .
https://github.com/MarcusLee-awsome/midi_file_experiment.git
this is my code's link  please give me some advice,please.

**Reply by apmcleod** [reply]
I ran your code and was getting the following exception:

```
No such file or directory: '../midi_degrader/data/degrade_midi_data/000.midi'
```

The issue here is that the directory `../midi_degrader/data/degrade_midi_data` (the `midi_path` variable in your code) does not exist. The fileio.df_to_midi function does automatically create missing subdirectories. You will need to create those folders first, manually or in the code, and then the code works (I tested this to confirm).

**Comment by MarcusLee-awsome** [comment]
thanks a lot. my computer can't run now.
give me a few minutes to share my code












李琦
***@***.***


---Original---
From: "James ***@***.***&gt;
Date: Thu, Oct 7, 2021 21:48 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [JamesOwers/midi_degradation_toolkit] Welcome to midi_degradation_toolkit Discussions! (Discussion #169)




 
I can try! Please share a minimal working code example showing what you’re trying to do which I can try to run.
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by MarcusLee-awsome** [comment]
Sorry，I want to bother you again. I want to know the details of the parameter about clean_prop and seed in the script of degrader.py. What factors should I use to confirm the value of this parameter. Please give some advice.

**Reply by apmcleod** [reply]
Hello,

`clean_prop` controls the proportion of calls to `Degrader.degrade` that will return a copy of the input without degrading it. If you want every excerpt to be degraded, for example, then you should set this value to 0.

`seed` just sets the seed for numpy's random number generator, which we use to randomly degrade each input. This is a standard argument for RNGs, which you can read more about, for example, [here](https://stats.stackexchange.com/questions/354373/what-exactly-is-a-seed-in-a-random-number-generator). Essentially, if you want every run of your script to produce exactly the same output given the same input, you should set this to any integer value. Otherwise, the default value of None will make every output differ randomly.

For more details, I can recommend looking at our [documentation notebooks](https://github.com/JamesOwers/midi_degradation_toolkit/tree/master/docs), including [number 4](https://github.com/JamesOwers/midi_degradation_toolkit/blob/master/docs/04_data_parsers_and_degrader.ipynb) which talks about the Degrader, and [number 3](https://github.com/JamesOwers/midi_degradation_toolkit/blob/master/docs/03_degradation_functions.ipynb) which talks about the degradation functions.

I would also recommend installing mdtk as described in the [readme](https://github.com/JamesOwers/midi_degradation_toolkit#install), rather than copying the source files into your code directory as you have done in the example you shared. This will ensure that all dependencies are installed correctly and that all code will work as intended.

**Comment by MarcusLee-awsome** [comment]
For example，if clean_prop = 1/3, only a music can be degraded in three picecs of music, and the proprtion of the type of degration is up to the degrations and degration_list. Am I right?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "JamesOwers/midi_degradation_toolkit"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月10日(星期天) 晚上7:31
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [JamesOwers/midi_degradation_toolkit] Welcome to midi_degradation_toolkit Discussions! (Discussion #169)





 
Hello,
 
clean_prop controls the proportion of calls to Degrader.degrade that will return a copy of the input without degrading it. If you want every excerpt to be degraded, then you should set this value to 0.
 
seed just sets the seed for numpy's random number generator, which we use to randomly degrade each input. This is a standard argument for RNGs, which you can read more about, for example, here.
 
For more details, I can recommend looking at our documentation notebooks, including number 4 which talks about the Degrader, and number 3 which talks about the degradation functions.
 
I would also recommend installing mdtk as described in the readme, rather than copying the source files into your code directory as you have done in the example you shared. This will ensure that all dependencies are installed correctly and that all code will work as intended.
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by apmcleod** [reply]
Not quite. An excerpt can be degraded an unlimited number of times. With `clean_prop=1/3`, every time you call Degrader.degrade, there is a 1/3 chance that the input will not be degraded. There is a 2/3 chance that the input will be degraded, according to the degradations.

Consider the following code snippet:
```
degrader = Degrader(clean_prop=1/3)
clean_count = 0

for i in range(1000):
    output_df, deg_id = degrader.degrade(input_df)
    if output_df.equals(input_df):
        clean_count += 1
```

After this runs, `clean_count` will be approximately 333.

However, all the degraded `output_df`s will be different.
